File: packages/wire/presence.go
package wire

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"errors"
	"strconv"
	"time"
)

// Remote presence and opaque rendezvous handle constants (V15-PR04).
const (
	DomainRendezvousHandle = "sendbeam/2 rendezvous-handle:"
	DomainPresenceProof    = "sendbeam/2 presence-proof:"

	DefaultRendezvousEpochWindow = 15 * time.Minute
	RendezvousHandleHexLength    = 64 // 32 bytes hex encoded
)

var (
	// ErrInvalidHandle indicates a malformed rendezvous handle.
	ErrInvalidHandle = errors.New("invalid rendezvous handle")

	// ErrInvalidProof indicates an invalid presence proof tag.
	ErrInvalidProof = errors.New("invalid presence proof")
)

// DeriveRendezvousHandle derives a 32-byte opaque rendezvous handle for a specific epoch index.
func DeriveRendezvousHandle(kPair []byte, epochIndex int64) string {
	epochStr := strconv.FormatInt(epochIndex, 10)
	mac := hmac.New(sha256.New, kPair)
	mac.Write([]byte(DomainRendezvousHandle))
	mac.Write([]byte(epochStr))
	return hex.EncodeToString(mac.Sum(nil))
}

// DeriveRendezvousHandleForTime derives the opaque handle for a given timestamp and window.
func DeriveRendezvousHandleForTime(kPair []byte, t time.Time, window time.Duration) string {
	if window <= 0 {
		window = DefaultRendezvousEpochWindow
	}
	epochIndex := t.UTC().Unix() / int64(window.Seconds())
	return DeriveRendezvousHandle(kPair, epochIndex)
}

// DeriveRendezvousHandlesWithSkew derives current, previous, and next epoch handles to tolerate clock drift.
func DeriveRendezvousHandlesWithSkew(kPair []byte, t time.Time, window time.Duration) []string {
	if window <= 0 {
		window = DefaultRendezvousEpochWindow
	}
	epochIndex := t.UTC().Unix() / int64(window.Seconds())
	return []string{
		DeriveRendezvousHandle(kPair, epochIndex-1),
		DeriveRendezvousHandle(kPair, epochIndex),
		DeriveRendezvousHandle(kPair, epochIndex+1),
	}
}

// DerivePresenceProof computes an HMAC proof of possession for registering or polling a handle.
func DerivePresenceProof(kPair []byte, handle string, nonce []byte) string {
	mac := hmac.New(sha256.New, kPair)
	mac.Write([]byte(DomainPresenceProof))
	mac.Write([]byte(handle))
	mac.Write(nonce)
	return hex.EncodeToString(mac.Sum(nil))
}

// VerifyPresenceProof validates an HMAC proof of possession in constant time.
func VerifyPresenceProof(kPair []byte, handle string, nonce []byte, proofHex string) bool {
	proofBytes, err := hex.DecodeString(proofHex)
	if err != nil || len(proofBytes) != sha256.Size {
		return false
	}
	expectedHex := DerivePresenceProof(kPair, handle, nonce)
	expectedBytes, _ := hex.DecodeString(expectedHex)
	return subtle.ConstantTimeCompare(proofBytes, expectedBytes) == 1
}

// ValidateRendezvousHandle checks whether a handle string is a valid 64-character lowercase hex string.
func ValidateRendezvousHandle(handle string) bool {
	if len(handle) != RendezvousHandleHexLength {
		return false
	}
	for i := 0; i < len(handle); i++ {
		c := handle[i]
		if (c < '0' || c > '9') && (c < 'a' || c > 'f') {
			return false
		}
	}
	return true
}

// MatchRendezvousHandle checks if a given handle matches any candidate epoch for k_pair.
func MatchRendezvousHandle(kPair []byte, handle string, t time.Time, window time.Duration) bool {
	if !ValidateRendezvousHandle(handle) {
		return false
	}
	candidates := DeriveRendezvousHandlesWithSkew(kPair, t, window)
	for _, cand := range candidates {
		if subtle.ConstantTimeCompare([]byte(cand), []byte(handle)) == 1 {
			return true
		}
	}
	return false
}

File: packages/protocol/src/presence.ts
/**
 * Remote presence handles, opaque rendezvous discovery, and blinded LAN beacon tags (V15-PR04).
 *
 * Matches Go `packages/wire/presence.go` and `packages/wire/lan_beacon.go` byte-for-byte.
 */

import { bytesToHex, concatBytes, utf8 } from './bytes.js';
import { hmacSha256 } from './webcrypto.js';

export const DOMAIN_RENDEZVOUS_HANDLE = 'sendbeam/2 rendezvous-handle:';
export const DOMAIN_PRESENCE_PROOF = 'sendbeam/2 presence-proof:';
export const DOMAIN_LAN_BEACON_TAG = 'sendbeam/2 lan-beacon:';

export const DEFAULT_RENDEZVOUS_EPOCH_WINDOW_MS = 15 * 60 * 1000; // 15 minutes
export const LAN_BEACON_TAG_SIZE = 16;
export const LAN_BEACON_NONCE_SIZE = 16;

/**
 * Constant-time hex string comparison.
 */
function constantTimeHexEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return diff === 0;
}

/**
 * Derive a 32-byte opaque rendezvous handle for a specific epoch index.
 */
export async function deriveRendezvousHandle(
  kPair: Uint8Array,
  epochIndex: number | bigint,
): Promise<string> {
  const epochStr = epochIndex.toString();
  const data = concatBytes(utf8(DOMAIN_RENDEZVOUS_HANDLE), utf8(epochStr));
  const tag = await hmacSha256(kPair, data);
  return bytesToHex(tag);
}

/**
 * Derive the opaque handle for a given timestamp and window.
 */
export async function deriveRendezvousHandleForTime(
  kPair: Uint8Array,
  t?: Date | number,
  windowMs = DEFAULT_RENDEZVOUS_EPOCH_WINDOW_MS,
): Promise<string> {
  const timeMs = t instanceof Date ? t.getTime() : typeof t === 'number' ? t : Date.now();
  const epochIndex = Math.floor(timeMs / windowMs);
  return deriveRendezvousHandle(kPair, epochIndex);
}

/**
 * Derive current, previous, and next epoch handles to tolerate clock drift.
 */
export async function deriveRendezvousHandlesWithSkew(
  kPair: Uint8Array,
  t?: Date | number,
  windowMs = DEFAULT_RENDEZVOUS_EPOCH_WINDOW_MS,
): Promise<string[]> {
  const timeMs = t instanceof Date ? t.getTime() : typeof t === 'number' ? t : Date.now();
  const epochIndex = Math.floor(timeMs / windowMs);
  return Promise.all([
    deriveRendezvousHandle(kPair, epochIndex - 1),
    deriveRendezvousHandle(kPair, epochIndex),
    deriveRendezvousHandle(kPair, epochIndex + 1),
  ]);
}

/**
 * Compute an HMAC proof of possession for registering or polling a handle.
 */
export async function derivePresenceProof(
  kPair: Uint8Array,
  handle: string,
  nonce: Uint8Array,
): Promise<string> {
  const data = concatBytes(utf8(DOMAIN_PRESENCE_PROOF), utf8(handle), nonce);
  const tag = await hmacSha256(kPair, data);
  return bytesToHex(tag);
}

/**
 * Verify an HMAC proof of possession in constant time.
 */
export async function verifyPresenceProof(
  kPair: Uint8Array,
  handle: string,
  nonce: Uint8Array,
  proofHex: string,
): Promise<boolean> {
  const expected = await derivePresenceProof(kPair, handle, nonce);
  return constantTimeHexEqual(proofHex.toLowerCase(), expected.toLowerCase());
}

/**
 * Derive a 16-byte truncated blinded tag for a paired device in a LAN beacon.
 */
export async function deriveLanBeaconTag(
  kPair: Uint8Array,
  nonce: Uint8Array,
  epochIndex: number | bigint,
): Promise<Uint8Array> {
  const epochStr = epochIndex.toString();
  const data = concatBytes(utf8(DOMAIN_LAN_BEACON_TAG), nonce, utf8(epochStr));
  const full = await hmacSha256(kPair, data);
  return full.slice(0, LAN_BEACON_TAG_SIZE);
}

/**
 * Compute candidate LAN beacon tags for [epoch-1, epoch, epoch+1].
 */
export async function deriveLanBeaconTagsForDevice(
  kPair: Uint8Array,
  nonce: Uint8Array,
  t?: Date | number,
  windowMs = DEFAULT_RENDEZVOUS_EPOCH_WINDOW_MS,
): Promise<Uint8Array[]> {
  const timeMs = t instanceof Date ? t.getTime() : typeof t === 'number' ? t : Date.now();
  const epochIndex = Math.floor(timeMs / windowMs);
  return Promise.all([
    deriveLanBeaconTag(kPair, nonce, epochIndex - 1),
    deriveLanBeaconTag(kPair, nonce, epochIndex),
    deriveLanBeaconTag(kPair, nonce, epochIndex + 1),
  ]);
}

/**
 * Match a candidate beacon against known paired device secrets.
 */
export async function matchLanBeaconTag(
  kPair: Uint8Array,
  beaconNonce: Uint8Array,
  advertisedTag: Uint8Array,
  beaconTimestampMs: number,
  nowMs = Date.now(),
  windowMs = DEFAULT_RENDEZVOUS_EPOCH_WINDOW_MS,
): Promise<boolean> {
  const skew = Math.abs(nowMs - beaconTimestampMs);
  if (skew > 2 * windowMs) {
    return false;
  }
  const candidates = await deriveLanBeaconTagsForDevice(
    kPair,
    beaconNonce,
    beaconTimestampMs,
    windowMs,
  );
  const advHex = bytesToHex(advertisedTag);
  for (const cand of candidates) {
    if (constantTimeHexEqual(bytesToHex(cand), advHex)) {
      return true;
    }
  }
  return false;
}
